learning-from-errors/overlapping-instances/article.md

```diff
@@ -0,0 +1,626 @@
+and `Printable (f a)` match.
+
+The next step says:
+
+```
+If no candidates remain, the search fails.
+```
+
+Since we have two instances that match, we can continue to the next step, which says,
+
+```
+Eliminate any candidate `IX` for which there is another candidate `IY` such that
+both of the following hold: `IY` is strictly more specific than `IX`. That is, `IY`
+is a substitution instance of `IX` but not vice versa. Either `IX` is overlappable,
+or `IY` is overlapping. (This “either/or” design, rather than a “both/and”
+design, allow a client to deliberately override an instance from a library
+without requiring a change to the library.)
+```
+
+We have two candidates, `Printable a` and `Printable (f a)`. Let us process
+`Printable a` first. The rule says to check if there is another candidate `IY`
+such that `IY` is a substitution instance of `IX`. Here `f a` is a
+substituation instance for `a`, because if something can accept `a`, it can accept
+`f a` as well, but not the other way around. So it fits, and the next part of the rule
+says that either `Printable a` is overlappable, or `Printable (f a)` should
+be overlapping. And since this does not match, we cannot eliminate `Printable a`.
+
+The next one is `Printable (f a)`, and we cannot eliminate it since the other
+instance `a` is not a substitution instance of `f a`. If something is expecting `f a`
+you cannot give `a` to it. Or in other words, `a` is not more specific than `f a`, but
+instead, it is more general.
+
+So after this rule, both instances remain, and the next rule says:
+
+```
+If all the remaining candidates are incoherent, the search succeeds, returning an arbitrary surviving candidate.
+```
+
+None of our instances are marked `Incoherent`, so we proceed to the next rule.
+
+```
+If more than one non-incoherent candidate remains, the search fails.
+```
+
+Considering that we have two such instances now, the lookup fails here.
+
+Let us try adding an `OVERLAPPING` pragma to the instance for `f a`.
+
+```
+instance {-# OVERLAPPING #-} Functor f => Printable (f a) where
+  printMe _ = putStrLn "Instance for a Functor"
+```
+
+And now we get the error:
+
+```
+• No instance for (Functor MyType) arising from a use of ‘printMe’
+• In the expression: printMe (MyType 'c')
+  In an equation for ‘main’: main = printMe (MyType 'c')
+```
+
+We can see that adding `OVERLAPPING` pragma enabled the elimination of the
+instance for `Printable a` at step 3. But the remaining instance `Functor f => Printable (f a)`
+failed to work, but this failure happens at a later phase: when constraints are matched and
+after GHC has picked an instance. This is why we get a `No instance for Functor MyType`
+error instead of an overlapping instance error.
+
+Here we come across a crucial aspect of instance resolution: the algorithm
+never backtracks. When the algorithm failed during matching the constraints, if it could
+backtrack, it could have picked the instance for `Printable a`, which was
```

Comment:
There’s another way to phrase this, which you could use even before going through the algorithm.

Instance selection looks at the **instance head** only, and never at the **instance context**. So when instance search takes place, GHC does not see candidate instances like this:

```
instance Printable a where
instance Functor f => Printable (f a)
```

It looks at them like this:

```
instance Printable a where
instance ... => Printable (f a)
```

And it has to pick an instance based on this information alone.

Reply:
Have included this description.